subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/transform/DefaultArtifactTransforms.java

```diff
@@ -0,0 +1,219 @@
+import org.gradle.api.internal.attributes.AttributeContainerInternal;
+import org.gradle.api.tasks.TaskDependency;
+import org.gradle.internal.component.local.model.ComponentFileArtifactIdentifier;
+import org.gradle.internal.component.model.DefaultIvyArtifactName;
+import org.gradle.internal.component.model.IvyArtifactName;
+import org.gradle.internal.resolve.ArtifactResolveException;
+
+import java.io.File;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+
+public class DefaultArtifactTransforms implements ArtifactTransforms {
+
+    private final ArtifactTransformRegistrationsInternal artifactTransformRegistrations;
+
+    public DefaultArtifactTransforms(ArtifactTransformRegistrations artifactTransformRegistrations) {
+        this.artifactTransformRegistrations = (ArtifactTransformRegistrationsInternal) artifactTransformRegistrations;
+    }
+
+    /**
+     * Returns a selector that selects the variant matching the supplied attributes, or which can be transformed to match. The selector may return null to mean 'none of these'
+     */
+    public <T extends HasAttributes> Transformer<T, Collection<? extends T>> variantSelector(final AttributeContainerInternal attributes) {
+        return new AttributeMatchingVariantSelector<T>(attributes.asImmutable());
+    }
+
+    /**
+     * Returns a visitor that transforms files and artifacts to match the requested attributes
+     * and then forwards the results to the given visitor.
+     */
+    public ArtifactVisitor visitor(final ArtifactVisitor visitor, @Nullable AttributeContainerInternal attributes) {
+        if (attributes == null || attributes.isEmpty()) {
+            return visitor;
+        }
+        return new ArtifactTransformingVisitor(visitor, attributes.asImmutable());
+    }
+
+    private class AttributeMatchingVariantSelector<T extends HasAttributes> implements Transformer<T, Collection<? extends T>> {
+        private final AttributeContainer attributes;
+
+        private AttributeMatchingVariantSelector(AttributeContainer attributes) {
+            this.attributes = attributes;
+        }
+
+        @Override
+        public String toString() {
+            return "Variant selector for " + attributes;
+        }
+
+        @Override
+        public T transform(Collection<? extends T> variants) {
+            // Note: This algorithm is a placeholder only. Should deal with ambiguous matches
+            if (attributes.isEmpty()) {
+                return variants.iterator().next();
+            }
+
+            // Note: This algorithm is a placeholder only. Should deal with ambiguous matches
+            T canTransform = null;
+            for (T variant : variants) {
+                AttributeContainer variantAttributes = ((AttributeContainerInternal) variant.getAttributes()).asImmutable();
+                if (artifactTransformRegistrations.areMatchingAttributes(variantAttributes, this.attributes)) {
```

Comment:
We shouldn't be using the registry as a cache here, but putting the cache into this class.